### PR TITLE
[Operator Development] fractional_max_pool2d & fractional_max_pool2d_backward

### LINKED ISSUE #1119: Code Contribution: 【Medium-Hard】【Operator Development】fractional_max_pool2d & fractional_max_pool2d_backward
## Description 任务介绍

Develop forward and backward functions for fractional_max_pool2d operator family.  
开发 fractional_max_pool2d 算子的前向和反向功能。

涉及两个算子：
- `fractional_max_pool2d`
- `fractional_max_pool2d_backward`

该算子属于 pooling 类算子，输出大小由随机采样控制。

## Requirements 任务要求

__Interface 接口__

Forward:  
`fractional_max_pool2d(Tensor self, int[2] kernel_size, int[2] output_size, Tensor random_samples) -> (Tensor, Tensor)`

Backward:  
`fractional_max_pool2d_backward(Tensor grad_output, Tensor self, int[2] kernel_size, int[2] output_size, Tensor indices) -> Tensor`

__Function reference 功能参考__

PyTorch 文档：  
https://pytorch.org/docs/stable/generated/torch.nn.FractionalMaxPool2d.html

__Implementation reference 实现参考__

PyTorch ATen 实现（行为参考）：  
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/FractionalMaxPool2d.cpp

FlagGems pooling 示例参考：  
https://github.com/FlagOpen/FlagGems/blob/master/src/flag_gems/ops/max_pool2d.py

__Behavior 行为要求__

- 行为必须与 PyTorch 完全一致：
  - 池化窗口根据 `kernel_size` 与 `random_samples` 生成随机偏移
  - 输出大小由 `output_size` 控制
  - 返回 `(output, indices)`，indices 用于 backward

- backward 需正确计算：  
  - `∂L/∂self`，并使用 forward 中的 indices 进行梯度回溯

- 支持所有参数，并确保所有 shape/dtype 校验行为与 PyTorch 对齐。

- `random_samples` 的 batch 维度必须与输入匹配。

## DDL 提交时间

请于接取任务后两周内提交 PR。  
请提供正确性测试与性能测试代码。

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by CarefreeSun**
### PR Category

Operator

### Type of Change

New Feature

### Description

Code Changes
- ```src/flag_gems/ops/fractional_max_pool2d.py``` - Core operator implementation
- ```src/flag_gems/ops/__init__.py``` - Export registration
- ```src/flag_gems/__init__.py``` - FlagGems operator registration
- ```tests/test_reduction_ops.py``` - Correctness tests
- ```benchmark/test_reduction_perf.py``` - Performance benchmarks

Operator Implementation

- The fractional_max_pool2d_forward_kernel dynamically calculates pooling windows using random_samples and scaling factors (alpha_h, alpha_w) to identify maximum values and cache their spatial indices.
- The backward pass is handled by fractional_max_pool2d_backward_kernel, which retrieves these stored indices to map gradients and uses tl.atomic_add to accurately accumulate values into grad_input.

Test Cases

The test cases are designed to comprehensively cover a diverse range of input, output, and kernel geometries - ranging from standard to extreme configurations like non-square, asymmetric, odd/even, or edge-case size - while also validating various batch/channel scales and FP16/FP32 data precisions.

### Issue

Resolves #1119

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### Performance

The test command lines and corresponding outputs are listed below.

```bash
pytest tests/test_reduction_ops.py -m fractional_max_pool2d
```
tests_for

```bash
pytest tests/test_reduction_ops.py -m fractional_max_pool2d_backward
```
tests_back

```bash
pytest benchmark/test_reduction_perf.py -m fractional_max_pool2d -s
```
屏幕截图 2026-01-06 192903

```bash
pytest benchmark/test_reduction_perf.py -m fractional_max_pool2d_backward -s
```
屏幕截图 2026-01-06 193004

### CODE REVIEW

**Comment by bin913** [comment]
@CarefreeSun  Please resolve conflicts

**Comment by bin913** [comment]
@CarefreeSun  Is this PR still needed? Can it be closed?

**Comment by douxetpur** [comment]
Hi, thank you for your participation in the FlagOS Open Computing Competition 🙏

After review, we have chosen to move forward with a different implementation for this operator. As a result, this PR will be closed.

We appreciate your time and contribution, and hope to see more of your submissions in the future.
